diff --git a/CMakeLists.txt b/CMakeLists.txt
index 89fe9f8a..a8126072 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,6 +46,14 @@ add_definitions(-DSQLITE_OMIT_LOAD_EXTENSION        # This one's for SQLiteCpp
                 -DSQLITE_ENABLE_FTS3_PARENTHESIS
                 -DSQLITE_ENABLE_FTS3_TOKENIZER)

+if(DEFINED SSL_LIB AND DEFINED CRYPTO_LIB)
+    if("${SSL_LIB}" STREQUAL "" OR "${CRYPTO_LIB}" STREQUAL "")
+        add_definitions(-DNO_SSL)
+    else()
+        add_definitions(-DOPENSSL_API_1_1 -DSSL_LIB=${SSL_LIB} -DCRYPTO_LIB=${CRYPTO_LIB})
+    endif()
+endif()
+
 if(BUILD_ENTERPRISE)
     add_definitions(-DCOUCHBASE_ENTERPRISE      # Tells LiteCore it's an EE build
                     -DSQLITE_HAS_CODEC          # Enables SQLite encryption extension (SEE)
diff --git a/Replicator/CivetWebSocket.cc b/Replicator/CivetWebSocket.cc
index b353b185..6c3c562c 100644
--- a/Replicator/CivetWebSocket.cc
+++ b/Replicator/CivetWebSocket.cc
@@ -120,7 +120,7 @@ namespace litecore { namespace websocket {
         ,_address(to)
         ,_options(options)
         {
-            mg_init_library(0);
+            mg_init_library(2);
         }

         ~CivetWebSocket() {
diff --git a/Replicator/Puller.cc b/Replicator/Puller.cc
index dda6761a..8df047cb 100644
--- a/Replicator/Puller.cc
+++ b/Replicator/Puller.cc
@@ -295,13 +295,14 @@ namespace litecore { namespace repl {
     
     Worker::ActivityLevel Puller::computeActivityLevel() const {
         ActivityLevel level;
-        if (_fatalError) {
+        if (_activeIncomingRevs > 0) {
+            level = kC4Busy;
+        } else if (_fatalError || !connected()) {
             level = kC4Stopped;
         } else if (Worker::computeActivityLevel() == kC4Busy
                 || (!_caughtUp && nonPassive())
                 || _waitingForChangesCallback
-                || _pendingRevMessages > 0
-                || _activeIncomingRevs > 0) {
+                || _pendingRevMessages > 0) {
             level = kC4Busy;
         } else if (_options.pull == kC4Continuous || isOpenServer()) {
             const_cast<Puller*>(this)->_spareIncomingRevs.clear();
diff --git a/Replicator/Pusher.cc b/Replicator/Pusher.cc
index dd349dac..c397c216 100644
--- a/Replicator/Pusher.cc
+++ b/Replicator/Pusher.cc
@@ -595,7 +595,9 @@ namespace litecore { namespace repl {
 
     Worker::ActivityLevel Pusher::computeActivityLevel() const {
         ActivityLevel level;
-        if (Worker::computeActivityLevel() == kC4Busy
+        if (!connected()) {
+            level = kC4Stopped;
+        } else if (Worker::computeActivityLevel() == kC4Busy
                 || (_started && !_caughtUp)
                 || _changeListsInFlight > 0
                 || _revisionsInFlight > 0
diff --git a/Replicator/Replicator.cc b/Replicator/Replicator.cc
index 0bd8f4c4..6a338d17 100644
--- a/Replicator/Replicator.cc
+++ b/Replicator/Replicator.cc
@@ -174,7 +174,7 @@ namespace litecore { namespace repl {
             case Connection::kDisconnected:
             case Connection::kClosed:
                 // After connection closes, remain active while I wait for db to finish writes:
-                level = (_dbStatus.level == kC4Busy) ? kC4Busy : kC4Stopped;
+                level = (max(max(Worker::computeActivityLevel(), _dbStatus.level), max(_pushStatus.level, _pullStatus.level)) == kC4Busy) ? kC4Busy : kC4Stopped;
                 break;
         }
         if (SyncBusyLog.effectiveLevel() <= LogLevel::Info) {
diff --git a/Replicator/Worker.hh b/Replicator/Worker.hh
index 5fed4793..a08a4401 100644
--- a/Replicator/Worker.hh
+++ b/Replicator/Worker.hh
@@ -157,6 +157,7 @@ namespace litecore { namespace repl {
 #if !DEBUG
     protected:
 #endif
+        bool connected() const                              {return _connection != nullptr;}
         blip::Connection* connection() const                {return _connection;}
 
     protected:
