diff --git a/CMakeLists.txt b/CMakeLists.txt
index 89fe9f8a..38761305 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,6 +46,10 @@ add_definitions(-DSQLITE_OMIT_LOAD_EXTENSION        # This one's for SQLiteCpp
                 -DSQLITE_ENABLE_FTS3_PARENTHESIS
                 -DSQLITE_ENABLE_FTS3_TOKENIZER)

+if(DEFINED SSL_LIB AND DEFINED CRYPTO_LIB)
+    add_definitions(-DOPENSSL_API_1_1 -DSSL_LIB=${SSL_LIB} -DCRYPTO_LIB=${CRYPTO_LIB})
+endif()
+
 if(BUILD_ENTERPRISE)
     add_definitions(-DCOUCHBASE_ENTERPRISE      # Tells LiteCore it's an EE build
                     -DSQLITE_HAS_CODEC          # Enables SQLite encryption extension (SEE)
diff --git a/Replicator/CivetWebSocket.cc b/Replicator/CivetWebSocket.cc
index b353b185..6c3c562c 100644
--- a/Replicator/CivetWebSocket.cc
+++ b/Replicator/CivetWebSocket.cc
@@ -120,7 +120,7 @@ namespace litecore { namespace websocket {
         ,_address(to)
         ,_options(options)
         {
-            mg_init_library(0);
+            mg_init_library(2);
         }

         ~CivetWebSocket() {
